### PR TITLE
Sharded tar writers for Lhotse Shar format

### PR DESCRIPTION
**Pull request by pzelasko**
This is part of the work needed to implement #846 

To make it easier to refer to this new data format I decided to name it `shar` (https://en.wikipedia.org/wiki/Lhotse_Shar), which can also be interpreted as "sharded archives".

I added tar writers that support saving recordings, features, arrays, and temporal arrays in a sharded manner. There is a high level `SharWriter` that shards the cutset and all of its relevant data files. It creates relevant writer for all data fields (supports wav/flac/mp3 for audio and numpy/lilcom for arrays). We remove any references to data in the cutset itself -- that data will be re-populated later by combining cutset+multiple tar files in a special reader class, and wrapped into `CutSet.from_shar` or something like that. I am also considering an option where it's no longer a CutSet, but I haven't decided yet.

### CODE REVIEW

**Comment by danpovey** [comment]
Apparently Lhotse Shar is the most deadly of the 8-thousander peaks, even worse than k2 :-(

**Comment by pzelasko** [comment]
That brings a new meaning to the expression "I will die on this hill" :)

**Comment by pzelasko** [comment]
I wrote two versions of readers: one uses ``torchdata``, the other doesn't use any external dependency. Also, I added some documentation for the code written so far. 

Technically this works now, but to be usable it requires at least the following features:
- [ ] Support for multi-worker and multi-node training, either one (or both) of these variants:
    - Data/shard de-duplication for DataLoader (WebDataset approach)
    - Different (but deterministic) shuffle of shards for every node+worker (Dan's suggestion; will make each epoch artificially longer by the factor of num_workers*num_gpus, but as Dan noted the user can select to checkpoint every N steps instead)

In addition to be more useful, it needs:
- [ ] Support for non-local filesystem storage (e.g. cloud)
- [ ] Support for reading just a subset of fields (as an option in readers)
- [ ] Support for extending an existing Shar archive with a new field (as an option in the writer or as a separate writer)
- [ ] Benchmarking vs WebDataset/single tar archive on slow storage so see how much behind it is due to multiple archive

Then, I would like to investigate how to extend/adapt this work to support:
- [ ] Advanced dataloader checkpointing for resuming training by skipping shards
- [ ] (maybe) ensuring equal number of batches from each dataloader worker

File: lhotse/audio.py [hunk 1]
```diff
@@ -470,13 +475,7 @@ def move_to_memory(
...
-            "wav",
-            "ogg",
-            "vorbis",
-            "flac",
-            "sph",
-        ):
+        if torchaudio_soundfile_supports_format() and format == "flac":
             # Prefer saving with soundfile backend whenever possible to avoid issue:
             # https://github.com/pytorch/audio/issues/2662
             # Saving with sox_io backend to FLAC may corrupt the file, IDK about other
```
Comment (desh2608) [inline lhotse/audio.py:481]: So, other formats (except flac) can be safely saved with the sox_io backend?

Comment (pzelasko) [reply lhotse/audio.py:481]: I believe so, at least I never experienced an issue with a different format.

**Comment by pzelasko** [comment]
I did a basic benchmark reading a single WebDataset tar file, and a collection of Shar archives from S3, and achieved ~3x greater throughput with Shar. This is probably because it could read the data in parallel using multiple connections, as it was using pipe input to read from `pipe:aws s3 cp <url> -`. I don't have any magnetic HDD around to test it with that, but I expect Shar at least shouldn't be (much) worse than WDS. I'll merge this as is with all tests passing. I might tweak some things later, and will try to figure out how to implement the training checkpointing mechanism.
